env/env_encryption.cc

```diff
@@ -361,9 +413,152 @@ Status EncryptedRandomRWFile::Close() { return file_->Close(); }
 
 // EncryptedEnv implements an Env wrapper that adds encryption to files stored
 // on disk.
-class EncryptedEnv : public EnvWrapper {
+class EncryptedEnvImpl : public EnvWrapper {
+  Status GetWritableProvider(const std::string& /*fname*/,
+                             EncryptionProvider** result) {
```

Comment:
seems like a "protocol violation".  Should this be a std::shared_ptr<EncryptionProvider> & result ?

Reply:
It was meant to be a raw pointer and not a shared (or other pointer).  First, the returned value is not "owned" by the caller.  It is a de-reference of an object within this class.  And the callers only need access to the underlying pointer, and should not care how it is memory managed.

Second, the raw pointer would allow the EnvEncryption to store the underlying provider via different means (raw/shared/unique pointer) without changing this interface/method or any of the callers.  

I will add a comment to outline what is returned (and why).  If you feel strongly that it should be a std::shared_ptr *, I can change it, but would prefer to leave it "as written"